### PR TITLE
chapel 1.24.0

### PR DESCRIPTION
**Pull request by Maxrimus**
Created with `brew bump-formula-pr`.

### CODE REVIEW

File: Formula/chapel.rb [hunk 1]
```diff
@@ -1,8 +1,8 @@
 class Chapel < Formula
   desc "Emerging programming language designed for parallel computing"
   homepage "https://chapel-lang.org/"
-  url "https://github.com/chapel-lang/chapel/releases/download/1.23.0/chapel-1.23.0.tar.gz"
-  sha256 "7ae2c8f17a7b98ac68378e94a842cf16d4ab0bcfeabc0fee5ab4aaa07b205661"
+  url "https://github.com/chapel-lang/chapel/releases/tag/1.24.0"
```
Comment (SMillerDev) [inline Formula/chapel.rb:4]: This is the release website, not the actual release.
```suggestion
  url "https://github.com/chapel-lang/chapel/releases/download/1.24.0/chapel-1.24.0.tar.gz"
```

**Comment by chenrui333** [comment]
```
python3 version 3.5 or later is required to install chpldoc and start_test dependencies. See https://www.python.org/
make[2]: *** [/usr/local/Cellar/chapel/1.24.0/libexec/third-party/chpl-venv/build/build-venv/ok] Error 1
make[1]: *** [chpldoc-venv] Error 2
make: *** [third-party-chpldoc-venv] Error 2
```

**Comment by bradcray** [comment]
Thanks for the commit @chenrui333.  I had gotten to the diagnosis, but not yet figured out how to address it in the formula.  Thanks for saving us some time there (with fingers crossed that this round of checks will pass).

**Comment by carlocab** [comment]
It looks like this formula is bundling a whole slew of libraries: https://github.com/chapel-lang/chapel/tree/master/third-party

It should be using Homebrew's versions of these where available instead.

**Comment by bradcray** [comment]
@carlocab : I understand your point and agree that this would be an improvement to the Chapel formula.

For historical context, we've done this bundling of third-party packages primarily for the convenience of our non-Homebrew users who build Chapel from our source tarball, often on clusters or supercomputers, where they can choose at compile-time between using their system installation of a package or the bundled version.  My impression is that when the initial version of our Homebrew formula was developed (by a community contributor outside of our team), this reliance on the bundled versions was inherited, either intentionally or unwittingly.  Since then, though we've taken over the updates to the formula as part of our semi-annual release process, we've essentially just bumped things along fairly mechanically every six months.  As far as I can recall, your comment is the first time that this concern has been flagged.

I think the right thing to do here is to survey our bundled third-party packages to determine:
* which are irrelevant or unnecessary for the Homebrew release and should just be `rm`'d in the formula (e.g., gasnet, ...)
* which represent cases that we could express as a Homebrew dependence (e.g., gmp, ...)
* which are required and either sufficiently different from their Homebrew versions or simply not available in homebrew such that they warrant bundling (e.g., qthreads, ...)

That said, I'm fairly certain that this analysis and change will take more effort than we have available time today.

For that reason, I'm curious whether it would be possible for us to proceed with this PR for today, with the promise to work on this improvement to the formula in the coming days and weeks, in order to permit our existing homebrew users to upgrade to today's release today.  I don't believe that the 1.24 formula is significantly different from the 1.23 formula in this regard, so arguably things are no worse than they have been, even though improvement is clearly possible and attractive.

Hopefully it's obvious, but none of this is intended to be pushing back against your feedback or concern; simply trying to see whether there's a path to completing our release process today before addressing it.  Thanks for your consideration.

**Comment by carlocab** [comment]
I'm fine with this; thanks for considering this carefully. I would like to see an issue tracking this on your repo, though, to try to avoid this falling off the radar.

**Comment by bradcray** [comment]
@carlocab: Thanks for your flexibility on this.  I was actually writing up the issue after posting the comment, and was just coming back to post the link to it here: https://github.com/chapel-lang/chapel/issues/17428

**Review by carlocab — APPROVED** [review]
Thanks, @Maxrimus, @bradcray.

**Comment by carlocab** [comment]
As an aside, qthreads looks like a neat library. I may look into trying to add that to homebrew/core.

**Comment by bradcray** [comment]
Thanks for your help in getting this in today, @SMillerDev, @chenrui333, and @carlocab!

Not at all critical, but: Late in the process, I noticed the "no arm bottle" label.  Though we don't have nearly as much Chapel QA on arm systems as we do on x86, my suspicion is that it ought to "just work."  If anyone has pointers to references about what's required to extend a formula to support an arm bottle, that could be interesting future work for us to take on.

**Comment by carlocab** [comment]
I'm going to trigger a build. If it succeeds, then you have your ARM bottle. If it fails, we can try to have a look at the logs.

**Comment by carlocab** [comment]
```
/opt/homebrew/Cellar/chapel/1.24.0/libexec/third-party/qthread/qthread-src/src/fastcontext/asm.S:46:4: error: What kind of a Mac is this?
#  error What kind of a Mac is this?
   ^
```
Looks like qthread (or, at least, the version of it that you bundle) doesn't know what an ARM Mac is.

**Comment by carlocab** [comment]
This suggests another advantage of using dependencies provided by Homebrew, btw. I've recently fixed a few formulae that were bundling their dependencies (but shouldn't), except that the bundled versions didn't build, and the ones on Homebrew do.

**Comment by bradcray** [comment]
> Looks like qthread (or, at least, the version of it that you bundle) doesn't know what an ARM Mac is.

I was just seeing that as well...  I'm pretty sure we build Qthreads for ARM, but probably not with clang (I think all of our arm platforms are linux-based and use gcc in nightly testing).  Let me check with others on the team who are more Qthreads-savvy to make sure I've got that straight and see if they have suggestions here.

A big workaround would be to try setting `CHPL_TASKS=fifo` before doing the build to opt out of Qthreads altogether; but it'd be preferable not to rely on that if we could get Qthreads to work.

**Comment by carlocab** [comment]
The issue may not be coming from Clang. Building on non-Apple ARM is only suggestive of compatibility with Apple ARM, but doesn't guarantee it, as there are some differences (e.g. calling conventions). This is also why we don't have ghc on ARM, even though there are Linux ARM builds for it.

**Comment by bradcray** [comment]
Got it, thanks for that context.  My team confirmed my suspicion that our only arm experience has been on x86 systems with gcc and cce (Cray's compiler).

I think the shortest path to getting this bottle up and running would be to make the `CHPL_TASKS=fifo` setting to disable Qthreads, but I don't feel strongly about pursuing that given that my "maybe it'll just work" hope didn't come true.  I'd be inclined to shelve this question for the time being and return to it after resolving https://github.com/chapel-lang/chapel/issues/17428 and/or when someone requests an arm bottle from us.

**Pull request closed without merging**